walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,223 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.*; 
+import java.util.Comparator;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    ArrayList<TimeRange> mandatoryAttendeeAvailability = new ArrayList<>();
+    ArrayList<TimeRange> allAttendeeAvailability = new ArrayList<>();
+
+    //if no attendees, any time of day works 
+    if (request.getAttendees().size() == 0 && request.getOptionalAttendees().size() == 0) {
+      allAttendeeAvailability.add(TimeRange.WHOLE_DAY);
+      return allAttendeeAvailability;
+    } 
+
+    //if the meeting is scheduled for longer than a day, it cannot be scheduled
+    else if (request.getDuration() > 24*60) {
+      return allAttendeeAvailability;
+    } 
+
+    //if there are no meetings, any time of day works 
+    else if (events.size() == 0) {
+      allAttendeeAvailability.add(TimeRange.WHOLE_DAY);
+      return allAttendeeAvailability;
+    }
+
+    //otherwise, find all possible meeting times   
+    else {
+
+      //first find the availability for mandatory attendees only 
+      ArrayList<Event> mandatoryAttendeeMeetings = new ArrayList<>(events);
+    
+      //remove existing meetings that have no overlapping attendees (not a conflict)
+      int i = 0;
+      while (i < mandatoryAttendeeMeetings.size()) {
+        ArrayList<String> newMeetingMandatoryAttendees = new ArrayList<>(request.getAttendees());
+        ArrayList<String> existingMeetingAttendees = new ArrayList<>(mandatoryAttendeeMeetings.get(i).getAttendees());
+        newMeetingMandatoryAttendees.retainAll(existingMeetingAttendees); 
+        if (newMeetingMandatoryAttendees.size() == 0) {
+          mandatoryAttendeeMeetings.remove(i);
+          continue;
+        } else {
+          i++;
+        }
+      }
+
+      //copy only the time ranges from the meetings, and sort 
+      ArrayList<TimeRange> mandatoryAttendeeMeetingTimes = new ArrayList<>();
+      for (int b = 0; b < mandatoryAttendeeMeetings.size(); b++) {
+        mandatoryAttendeeMeetingTimes.add(mandatoryAttendeeMeetings.get(b).getWhen());
+      }
```

Comment:
This part seems a little overly complex. Instead of initializing `mandatoryAttendeeMeetings` with all the events, removing the events they're not a part of, and then copying each meeting's time range to another `mandatoryAttendeeMeetingTimes` array, you could just loop through the events, check if the mandatory attendees are a part of the event, and if so, then add that meeting time to `mandatoryAttendeeMeetingTimes`.

Reply:
Good point, this is definitely overly complex. I just have a quick question, if that's okay. The reason I copied the meetings from the collection events to an ArrayList was that I couldn't figure out a way to access the different attributes, like getAttendees() or getWhen() directly from the collection. I could iterate through it, using an Iterator, but I couldn't access the values (I looked up ways to access objects in a collection but didn't find anything that solved this problem, only how to access the entire object). Would it be okay to first copy the meetings to an ArrayList, and then simplify it like you said by looping through that ArrayList and directly adding the times to mandatoryAttendeeMeetingTimes if there are overlapping attendees?

Reply:
If you use an Iterator to iterate through `events`, you'd be able to get the next `Event` in the collection with `iterator.next()`. If you assign that to a variable (e.g. `Event event = iterator.next();`), you should be able to call getAttendees() or getWhen() on `event`. Alternatively, you could use a for each-loop: `for (Event event : events)`.

Reply:
Okay, thank you so much. I was trying to do it all in one step (interator.next().getAttendees()) instead of saving it as a variable first. Sorry I didn't figure that out sooner on my own!